test/test_acousticbrainz.py

```diff
@@ -0,0 +1,206 @@
+                              )
+
+    def test_check_not_duplicate_and_fail_extraction(self):
+        self.tagger.webservice.get = Mock(wraps=self.mock_get_fail)
+        ab_feature_extraction(self.tagger,
+                              self.file.metadata["musicbrainz_recordingid"],
+                              "fail.mp3",
+                              partial(self.mock_ab_extractor_callback_extraction_failed, self.tagger, self.file)
+                              )
+
+    def test_check_not_duplicate_and_succeed_extraction(self):
+        self.tagger.webservice.get = Mock(wraps=self.mock_get_fail)
+        ab_feature_extraction(self.tagger,
+                              self.file.metadata["musicbrainz_recordingid"],
+                              "test.mp3",
+                              partial(self.mock_ab_extractor_callback_extraction_succeeded, self.tagger, self.file)
+                              )
+
+
+class AcousticBrainzFeatureSubmissionTest(AcousticBrainzFeatureExtractionTest):
+    def setUp(self):
+        super().setUp()
+        self.tagger.webservice.get = Mock(wraps=self.mock_get_fail)
+
+        # Setup methods and settings required by file.set_pending and file.clear_pending
+        self.tagger.tagger_stats_changed = MagicMock()
+        self.tagger.tagger_stats_changed.emit = Mock(wraps=self.mock_emit)
+        self.set_config_values(settings)
+
+    @staticmethod
+    def mock_ab_submission_failed(tagger, file, result, error):
+        AcousticBrainzFeatureExtractionTest.singleton.assertEqual(error, None)
+        ab_metadata_file, result, error = result
+        AcousticBrainzFeatureExtractionTest.singleton.assertTrue("Duplicate" not in error or "Writing results" not in error)
+
+    @staticmethod
+    def mock_ab_submission_succeeded(tagger, file, result, error):
+        AcousticBrainzFeatureExtractionTest.singleton.assertEqual(error, None)
+        ab_metadata_file, result, error = result
+        AcousticBrainzFeatureExtractionTest.singleton.assertTrue("Writing results" in error)
+
+    def mock_emit(self):
+        pass
+
+    def mock_post_succeed(self, host, port, path, data, handler, parse_response_type, priority, important, mblogin, queryargs=None, request_mimetype=None):
+        handler("""{"message":"ok"}""", "", 0)
+
+    def mock_post_fail(self, host, port, path, data, handler, parse_response_type, priority, important, mblogin, queryargs=None, request_mimetype=None):
+        handler("""APIBadRequest""", "", 400)
+
+    def test_submit_failed(self):
+        self.file.set_pending()
+        self.tagger.webservice.post = Mock(wraps=self.mock_post_fail)
+        ab_feature_extraction(self.tagger,
+                              self.file.metadata["musicbrainz_recordingid"],
+                              "test.mp3",
+                              partial(ab_extractor_callback, self.tagger, self.file)
+                              )

```

Comment:
We usually split a bit differently, first argument is moved to next line and indented in order to reclaim room on right side:

```python
ab_feature_extraction(
    self.tagger,
    self.file.metadata["musicbrainz_recordingid"],
    "test.mp3",
    partial(ab_extractor_callback, self.tagger, self.file)
)
```